x/oracle/keeper/query_server_test.go

```diff
@@ -0,0 +1,129 @@
+		err      string
+	}{
+		{
+			name:     "failure - should handle nil request",
+			req:      nil,
+			expected: nil,
+			err:      "rpc error: code = InvalidArgument desc = invalid request",
+		},
+		{
+			name:     "success - should return correct oracle address",
+			req:      &types.QueryOracleAddressRequest{},
+			expected: &types.QueryOracleAddressResponse{Address: ""},
+		},
+		{
+			name:     "success - should return correct oracle address",
+			req:      &types.QueryOracleAddressRequest{},
+			oracle:   "cosmos1w6t0l7z0yerj49ehnqwqaayxqpe3u7e23edgma",
+			expected: &types.QueryOracleAddressResponse{Address: "cosmos1w6t0l7z0yerj49ehnqwqaayxqpe3u7e23edgma"},
+		},
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			if len(tc.oracle) > 0 {
+				s.app.OracleKeeper.SetOracle(s.ctx, sdk.MustAccAddressFromBech32(tc.oracle))
+			}
+			resp, err := s.app.OracleKeeper.OracleAddress(s.ctx, tc.req)
+			if len(tc.err) > 0 {
+				s.Assert().EqualError(err, tc.err, "should return the correct error")
+				s.Assert().Nil(resp, "response should be nil")
+			} else {
+				s.Assert().NoError(err, "should not return an error")
+				s.Assert().Equal(tc.expected, resp, "should return the correct response")
+			}
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestOracle() {
+	tests := []struct {
+		name        string
+		req         *types.QueryOracleRequest
+		expected    *types.QueryOracleResponse
+		oracle      string
+		mockEnabled bool
+		err         string
+	}{
+		{
+			name: "failure - should handle nil request",
+			req:  nil,
+			err:  "rpc error: code = InvalidArgument desc = invalid request",
+		},
+		{
+			name: "failure - should handle invalid query data",
+			req: &types.QueryOracleRequest{
+				Query: []byte("abc"),
+			},
+			err: "rpc error: code = InvalidArgument desc = invalid query data",
+		},
+		{
+			name: "failure - should handle unset oracle",
+			req: &types.QueryOracleRequest{
+				Query: []byte("{}"),
+			},
+			err: "missing oracle address",
+		},
+		{
+			name: "success - should handle error from contract",
+			req: &types.QueryOracleRequest{
+				Query: []byte("{}"),
+			},
+			err: "missing oracle address",
+		},
+		{
+			name: "failure - should handle error from contract",
+			req: &types.QueryOracleRequest{
+				Query: []byte("{}"),
+			},
+			oracle: "cosmos1w6t0l7z0yerj49ehnqwqaayxqpe3u7e23edgma",
+			err:    "contract: not found",
+		},
+		{
+			name: "success - should handle response from contract",
+			req: &types.QueryOracleRequest{
+				Query: []byte("{}"),
+			},
+			oracle: "cosmos1w6t0l7z0yerj49ehnqwqaayxqpe3u7e23edgma",
+			expected: &types.QueryOracleResponse{
+				Data: []byte("{}"),
+			},
+			mockEnabled: true,
+		},
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			if tc.mockEnabled {
+				s.app.OracleKeeper.SetWasmQueryServer(keeper.MockWasmServer{})
+			}
+
+			if len(tc.oracle) > 0 {
+				s.app.OracleKeeper.SetOracle(s.ctx, sdk.MustAccAddressFromBech32(tc.oracle))
+			}
```

Comment:
A word of caution on this: these lines are doing things that will affect the tests cases to follow.

There's two workarounds:
1. Change the setters into `With...` functions, create a keeper variable to hold the keeper to use for the test, and, if mocking, call the `With...` function and asign the result back to the keeper variable.
2. Before calling the setter, get the current thing and create a `defer` that sets it back to its original value.

Example of what `WithWasmQueryServer` would look like:
```golang
func (k Keeper) SetWasmQueryServer(server wasmtypes.QueryServer) Keeper {
	k.wasmQueryServer = server
	return k
}
```

Since it's got a concrete receiver (no `*`), it doesn't end up updating the existing keeper; it ultimately creates a copy of it with the new `wasmQueryServer` value.

Relatedly, with mocks, it's usually best to always mock if you're going to mock any test cases. The place where it might cause problems is if you've got a test that you expect to not need the mocked stuff, but suddenly it does. Now you're making those calls for real which might be causing other problems.

Reply:
I'm happy you mentioned this. This is exactly what I wanted to do.